src/Optimizer/Transformer/AmpRuntimePreloads.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace AmpProject\Optimizer\Transformer;
+
+use AmpProject\Amp;
+use AmpProject\Dom\Document;
+use AmpProject\Optimizer\ErrorCollection;
+use AmpProject\Optimizer\Transformer;
+use AmpProject\RequestDestination;
+
+/**
+ * Transformer adding resource hints to preload the AMP runtime script and CSS.
+ *
+ * @package ampproject/amp-toolbox
+ */
+final class AmpRuntimePreloads implements Transformer
+{
+    /**
+     * Apply transformations to the provided DOM document.
+     *
+     * @param Document        $document DOM document to apply the transformations to.
+     * @param ErrorCollection $errors   Collection of errors that are collected during transformation.
+     * @return void
+     */
+    public function transform(Document $document, ErrorCollection $errors)
+    {
+        if ($this->isAmpRuntimeScriptNeeded($document)) {
+            $document->links->addPreload($this->getAmpRuntimeScriptHost(), RequestDestination::SCRIPT);
+        }
```

Comment:
This is currently being handled by the `RewriteAmpUrls` transformer. Should the logic be moved here? At the very least it seems `RewriteAmpUrls` should make use of the `LinkManager` in its `createPreload` method.

Reply:
I made `RewriteAmpUrls` reuse the link manager. I also extended the link manager so that it ensure no duplicate links are being produced. This currently uses both `href` and `rel` for the uniqueness key.